inox2d/src/render.rs

```diff
@@ -1,378 +1,386 @@
+		self.begin_composite_content(as_mask, components, render_ctx, id);
 
-		for &uuid in children {
-			let node = puppet.nodes.get_node(uuid).unwrap();
-			let node_render_ctx = &puppet.render_ctx.node_render_ctxs[&uuid];
-
-			if let (InoxData::Part(ref part), RenderCtxKind::Part(ref part_render_ctx)) =
-				(&node.data, &node_render_ctx.kind)
-			{
-				if as_mask {
-					self.draw_part_self(true, camera, node_render_ctx, part, part_render_ctx);
-				} else {
-					self.draw_part(camera, node_render_ctx, part, part_render_ctx, puppet);
+		for uuid in &render_ctx.zsorted_children_list {
+			let drawable_kind =
+				DrawableKind::new(*uuid, comps).expect("All children in zsorted_children_list should be a Drawable.");
+			match drawable_kind {
+				DrawableKind::TexturedMesh(components) => {
+					self.draw_textured_mesh_content(as_mask, &components, comps.get(*uuid).unwrap(), *uuid)
 				}
-			} else {
-				// composite inside composite simply cannot happen
+				DrawableKind::Composite { .. } => panic!("Composite inside Composite not allowed."),
 			}
 		}
 
-		self.finish_composite_content(as_mask, comp);
+		self.finish_composite_content(as_mask, components, render_ctx, id);
 	}
 
-	fn draw(&self, camera: &Mat4, puppet: &Puppet) {
-		for &uuid in &puppet.render_ctx.root_drawables_zsorted {
-			let node = puppet.nodes.get_node(uuid).unwrap();
-			let node_render_ctx = &puppet.render_ctx.node_render_ctxs[&uuid];
-
-			match (&node.data, &node_render_ctx.kind) {
-				(InoxData::Part(ref part), RenderCtxKind::Part(ref part_render_ctx)) => {
-					self.draw_part(camera, node_render_ctx, part, part_render_ctx, puppet);
-				}
-
-				(InoxData::Composite(ref composite), RenderCtxKind::Composite(ref children)) => {
-					self.draw_composite(false, camera, composite, puppet, children);
-				}
-
-				_ => {
-					// This clearly is sign that the data structure needs rework
-					todo!();
-				}
-			}
+	fn draw(&self, puppet: &Puppet) {
+		for uuid in &puppet
+			.render_ctx
+			.as_ref()
+			.expect("RenderCtx of puppet must be initialized before calling draw().")
+			.root_drawables_zsorted
+		{
+			self.draw_drawable(false, &puppet.node_comps, *uuid);
 		}
 	}
 }
+
+/// Dispatches draw calls for all nodes of `puppet`
+/// - with provided renderer implementation,
+/// - in Inochi2D standard defined order.
+///
+/// This does not guarantee the display of a puppet on screen due to these possible reasons:
+/// - Only provided `InoxRenderer` method implementations are called.
+/// For example, maybe the caller still need to transfer content from a texture buffer to the screen surface buffer.
+/// - The provided `InoxRender` implementation is wrong.
+/// - `puppet` here does not belong to the `model` this `renderer` is initialized with. This will likely result in panics for non-existent node uuids.
+pub fn draw<T: InoxRenderer>(renderer: &T, puppet: &Puppet) {
+	renderer.on_begin_draw(puppet);
+	renderer.draw(puppet);
+	renderer.on_end_draw(puppet);
+}
```

Comment:
I don't think it's a good idea to define this function. Calling `on_begin_draw` then `draw` then `on_end_draw` is more easy to understand than calling a separate `inox2d::render::draw(...)` function which itself has to be inspected.